docassemble/ALToolbox/business_days.py

```diff
@@ -0,0 +1,87 @@
+import holidays  
+import pandas as pd
+import datetime 
+from datetime import date as dt  
+from docassemble.base.util import as_datetime
+"""
+  External docs: 
+  1. https://github.com/dr-prodigy/python-holidays
+  2. https://pypi.org/project/holidays/
+  3. https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html
+  Note:
+  There seems to be a core issue in holidays module, where if you delete a holiday cross the year, it will add in holidays from the previous year in error. For example, 
+      12/31/2021: New Year’s Day (Observed) and 01/01/2022: New Year’s Day 
+  are both included in 2022 holidays. If remove_holidays=['new year'], then after this line of code:
+      local_holidays.pop_named(record) 
+  the results will erroneously include 2021's holidays as well. But practically no one would delete New Year anyway, we should be fine and there is no need to alarm the user about this issue.
+"""
+def non_business_days(date, interval=0, country='US', state='MA', province=None, add_holidays=None, remove_holidays=None, first_n_dates=0) -> dict:      
+  # 1. Collect weekends, standard holidays and user defined add/remove holidays
+  # 1.1 Get all saturdays and sundays in the given year  
+  date = as_datetime(date)
+  year = date.year
```

Comment:
A related but different question to year boundaries is if we need to pass in the date at all? It's only using the year of the date, so would it make more sense to just directly pass the year (or years if we want to support generating multiple years of holidays?)

Reply:
Strictly speaking, you are right that only the year matters for this particular function. 

However, there are three functions in this module with similar inputs, for user's sake it might be easier to keep the input items consistent for all these functions, as described in the intro page. What do you think?

Reply:
I think either works fine: for the possible extension we discussed earlier, you can auto get holidays for the next year by looking at the date, so you're guaranteed holidays for 60 days out, or you could pass the years, so you always get all holidays for all the years you requested. I'll leave it up to your judgement, which seems to be changing it to year.

Reply:
I just discovered that `holidays `module had a major change about 2 weeks ago, the readme page changed, and the function I am using (CountryHoliday) is deprecated! What a perfect timing! https://github.com/dr-prodigy/python-holidays/commit/18f02630520ea7da29c86200ce8143cb6dac0a26. Oh well, better now than after we released our code. I'll make the switch.
![image](https://user-images.githubusercontent.com/20231649/154299247-11674bc2-86e5-45f9-8630-559495d804fb.png)

Reply:
On a positive note, I tested some cross-year scenarios using the current code and they worked. I think it's because we loop thru each day and call non_business_days for it individually, which means we are always dealing with a "current day", thus no worries about year boundary. 

I am glad we are having this discussion, it made me think how users will use these functions and what input items are relevant. Maybe I'll add a "holidays" function as well, because developers may need to check what's on the shelve in order to determine which ones to add/remove. They can also list holidays as an output if needed. So you'll see more changes coming 😀